balancer/rls/internal/control_channel.go

```diff
@@ -150,33 +140,34 @@ func (cc *controlChannel) monitorConnectivityState() {
 	// returning only one new picker, regardless of how many backoff timers are
 	// cancelled.
 
+	// Using the background context is fine here since we check for the ClientConn
+	// entering SHUTDOWN and return early in that case.
+	ctx := context.Background()
+
 	// Wait for the control channel to become READY.
-	for s := cc.cc.GetState(); s != connectivity.Ready; s = cc.cc.GetState() {
-		if !cc.cc.WaitForStateChange(cc.ctx, s) {
-			// cc.ctx has expired. Can happen only if close() was called.
-			return
-		}
+	if !cc.waitForReady(ctx) {
+		return
 	}
-	cc.logger.Infof("Connectivity state is %s", cc.cc.GetState())
+	cc.logger.Infof("Connectivity state is READY")
 
 	for {
-
-		// Wait for the control channel to enter anything other than READY.
-		state := cc.cc.GetState()
-		if !cc.cc.WaitForStateChange(cc.ctx, state) {
+		// There is a small chance that the control channel has moved out of READY
+		// since the time we last saw it enter READY. Wait for it to enter anything
+		// other than READY.
+		if !cc.cc.WaitForStateChange(ctx, connectivity.Ready) {
 			// cc.ctx has expired. Can happen only if close() was called.
 			return
 		}
+		if cc.cc.GetState() == connectivity.Shutdown {
+			return
+		}
 		cc.logger.Infof("Connectivity state is %s", cc.cc.GetState())
 
 		// Wait for the control channel to become READY again.
-		for s := cc.cc.GetState(); s != connectivity.Ready; s = cc.cc.GetState() {
-			if !cc.cc.WaitForStateChange(cc.ctx, s) {
-				// cc.ctx has expired. Can happen only if close() was called.
-				return
-			}
+		if !cc.waitForReady(ctx) {
```

Comment:
Another option: do this at the top of the loop, remove the separate function (since you don't need to do it twice anymore), and make pushing on the channel conditional on `!first` or something.  Totally optional; this is fine too.